Refactor and refine cascade deletion rules

melton-jason opened the pull request:
Fixes https://github.com/specify/specify7/issues/1694, https://github.com/specify/specify7/issues/2784, https://github.com/specify/specify7/issues/1069
Additionally test #2511 to see if the error still occurs. If so, we can easily add AddressofRecord to Borrow deletion blockers. 

In addition to refactoring the code, the following rules were added:

While deleting any of the following:
- Picklist should delete Picklist Item
- Spquery should delete Spqueryfield
- Spauditlog should delete Spauditlogfield
- Specifyuser should delete Spprincipal

In addition, we now have the ability to extend deletion blockers as we see fit on the backend. 
Added:
- Specifyuser added to the Agent delete blockers (You should not be able to delete an Agent when associated with Specifyuser)

melton-jason committed:
Refactor and fine tune deletion rules
Fixes #1694, #2784, #1069

melton-jason committed:
Add ability to extend deletion blockers

This was originally in commit 56cb36f2, but is being moved here
as this is more fitting for the functionality and easier to find
in the future

maxxxxxdlp commented on specifyweb/specify/build_models.py:
```diff
@@ -112,11 +96,13 @@ def make_relationship(modelname, rel, datamodel):
 
     try:
-        on_delete = SPECIAL_DELETION_RULES["%s.%s" % (modelname.capitalize(), rel.name.lower())]
+        on_delete = SPECIAL_DELETION_RULES[f"{rel.name.capitalize()}"][f"{modelname.lower()}"]
```
It looks like you swapped the relationship and model name. Is this intentional?

melton-jason replied:
I think I should change the variable names...

This was confusing even in production. 

Take this for example. 
https://github.com/specify/specify7/blob/741a5e34da50328ce8706d25c2071faba3976a2e/specifyweb/specify/build_models.py#L77
This operates as 'When a Specifyuser is deleted, set its relationship with Agent to null"
Where the on_delete functionality for these rules get defined, the `modelname` would be Agent, and `rel.name` would be `specifyuser`.
https://github.com/specify/specify7/blob/741a5e34da50328ce8706d25c2071faba3976a2e/specifyweb/specify/build_models.py#L115

I thought it might make more sense to have the table being deleted be the 'primary' key of the dictonary, so I changed it to `'Specifyuser' : {'agent': models.SET_NULL}`

As a result of this flip, I flipped the variables where `on_delete` was being defined to ensure the same functionality.

melton-jason replied:
I can change it to `'Agent': {'specifyuser': models.SET_NULL}` if that makes more sense

maxxxxxdlp replied:
I think adding an example breakdown you provided here as a comment on top of `SPECIAL_DELETION_RULES` (in addition to what you already have there) would be great

maxxxxxdlp commented on specifyweb/specify/deletion_rules.py:
```diff
@@ -0,0 +1,67 @@
+    
+    Of the form: base_table : { relationship: action}
+
+    Possible Django actions (w/ django 2.2) are:
```
Thanks for documenting this!

grantfitzsimmons committed:
Merge branch 'production' into issue-1694

grantfitzsimmons commented:
Add a block for
- `spdataset`
- Queries
- Reports & Labels (if they aren't already covered by `spappreosurce`)

melton-jason commented:
> Add a block for
> 
> * `spdataset`

For more context, this is regarding adding a deletion blocker to SpecifyUser. 
This operation is not currently supported with how I extended deletion blockers, but I will add a way to support adding reverse relationships so this can be added.

melton-jason commented:
> > Add a block for
> > 
> > * `spdataset`
> 
> For more context, this is regarding adding a deletion blocker to SpecifyUser. This operation is not currently supported with how I extended deletion blockers, but I will add a way to support adding reverse relationships so this can be added.

`Spatasets` will more complicated than any other table(s) to work with on the backend. Many of Specify's internal helper functions rely on the datamodel being loaded when Specify starts up (this happens in [load_datamodel.py](https://github.com/specify/specify7/blob/production/specifyweb/specify/load_datamodel.py)). 
The issue is that to get the list of tables and fields in the database, Specify 7 reads the `specify_datamodel.xml` in the Specify 6 config directory (as seen below), and as far as I know the spdataset table is not included within that file.
https://github.com/specify/specify7/blob/4877f75686668cdd5b2d28c46881fb7628ce69a4/specifyweb/specify/load_datamodel.py#L222-L231

We could manually use SQL to get the correct information, but that will only be a short-term solution (which may be fine for now). Preferably, we would want to engineer a way to load/register these Specify 7 specific tables in the internal specify datamodel. (If such a way already exists and I am overlooking it, then feel free to point it out. I have checked all attributes on specify.models and spdataset does not exist)

melton-jason commented:
Additionally requested by @grantfitzsimmons:
Add a Report/Label deletion blocker to Specifyuser

melton-jason committed:
Correct field names

melton-jason commented:
I have went through and cleaned up all of the deletion rules as they stand, so this is ready for testing. 

Adding reverse-relationship deletion blockers (i.e. the relationship we want to protect does not exist on the table we want to delete) will require a little more engineering and time: there is no reason to further delay the changes that have already been made on this branch. 

These deletion rules can probably be addressed/changed when #2847 and #2813 are being developed (once those two issues are being developed, #2584 can be fixed as well), but this branch should probably be merged before work on those issues begin. 

Also, along with testing #2511, can issues #1385, #1431, #88, and #150 additionally be tested to see if those issues still occur? If so, they can be fixed in this PR.

melton-jason commented:
| Base Table  | Relationship | Action |
| ------------- | ------------- | ------------- |
| Spappresource  | SpReports  | Cascade  |
| Recordset  | RecordSet Items  | Cascade  |
| Specifyuser | Agent  | Sets Null  |

melton-jason committed:
Delete 'names' rather than 'name' in splocalecontaineritem

Additionally remove lines from agent_rules.py
which were used for testing, and should not have
gotten pushed to origin!

grantfitzsimmons commented:
It looks like #1385 is not solved

> <img width="1280" alt="image" src="https://user-images.githubusercontent.com/37256050/213751802-c23eb789-fb2d-4602-9bc3-f01986e2f22b.png">
> 
> https://fwri2023-issue-1694.test.specifysystems.org/specify/query/48/
> 
> #2806 does not seem to help this

grantfitzsimmons commented:
Does not solve #1431 either

> Still able to recreate
> 
> https://fwri2023-issue-1694.test.specifysystems.org/specify/view/collectionobject/367315/
> 
> <img width="1273" alt="image" src="https://user-images.githubusercontent.com/37256050/213752479-1733a47f-7f1d-4709-b94b-fd13a4006537.png">
>

grantfitzsimmons commented:
I couldn't recreate #88 (likely because I do not understand the issue) and #150 (seems to be invalid/already solved).

grantfitzsimmons commented:
#2511 is not solved either

> https://herbrbge6802-issue-1694.test.specifysystems.org/specify/view/borrow/1365/
> 
> <img width="1253" alt="image" src="https://user-images.githubusercontent.com/37256050/213754302-cc9a2ddf-eaa0-4cb3-8038-21a496fc2248.png">
>

grantfitzsimmons reviewed (commented):
Please let me know if everything I discovered is intentional. I can formally aprove once I hear from you again

melton-jason committed:
Fix to-manys cascade deletion rules

Turns out the models contain the attribute name as plural,
 to signify this is a to-many relationship.

For deletion rules, the table name is needed and NOT the name
of the attribute on the base_table model

melton-jason committed:
Add typing to build_models.py

melton-jason committed:
Reformat special deletion rules

melton-jason committed:
Add remaining deletion rules

Fixes #2511, #1431

melton-jason commented:
These remaining issues should be fixed now.

- Spauditlogfield should no longer block the deletion of Spauditlog

- #1431
  - This will require some discussion. How should we handle LoanPreparation when it's associated Preparation is deleted?  For comparison purposes to other branches I have set the relationship to null between Preparation and LoanPreparation. (The data seems to be preserved within the LoanPreparation, but it is 'orphaned' )
  
  - On other branches, LoanPreparations are a deletion blocker for Preparations, but the 'Trash' icon can still be clicked from the CollectionObject form and the user can 'attempt' to remove the Preparation. If we want to preserve the deletion blocker, we would need an implementation in the front end

Some concerns:
- #2511 

Correctly solving this issue would be impossible with the current way the backend sets up its deletion rules (There is no `Addressofrecord.borrows` relationship). The code can be refactored to support handling cases like these, but I am not sure how bugprone the solution may be if not hardcoded. 

There is another concerning side-effect with this issue (currently in production). When Borrow is deleted, it tries to cascade and delete its associated Addressofrecord. If there are multiple records referencing the Addressofrecord, the error is thrown. However, if the Borrow is the only record referencing the Addressofrecord, both the Addressofrecord _and_ Borrow will be deleted. 
This actually holds true for all of the following relationships:

- Accession.addressofrecord
- Borrow.addressofrecord
- Exchangein.addressofrecord
- Exchangeout.addressofrecord
- Gift.addressofrecord
- Loan.addressofrecord
- Repositoryagreement.addressofrecord

Interactions are flagged as dependents of Addressofrecord (i.e., they require an Addressofrecord), but at a database level they do not require an Addressofrecord (similar to #2299). 
@maxxxxxdlp or @grantfitzsimmons  do you know why there is this discrepancy? 

- #2784 

Spprincipal and Specifyuser have a many-to-many relationship: a relationship type which is not created/supported on the backend. We would either need a proper implementation to handle creating Django many-to-many relationships (the on delete behavior would have to be [handled differently](https://stackoverflow.com/questions/6333068/django-removing-object-from-manytomany-relationship) than other fields) or use SQL directly to handle deletion rules on a case-by-case basis. 

- #1385 

To maintain compatibility with Specify 6, I will not touch the schema mapping fields in this pull request. This can be addressed again when the frontend gets a 'proper' implementation of schema mappings.

grantfitzsimmons commented:
@melton-jason I need to take a look at this again.

>Interactions are flagged as dependents of Addressofrecord (i.e., they require an Addressofrecord), but at a database level they do not require an Addressofrecord (similar to https://github.com/specify/specify7/issues/2299).
@maxxxxxdlp or @grantfitzsimmons do you know why there is this discrepancy?

I would remove that requirement– that's bizarre. What do you think @maxxxxxdlp ?

maxxxxxdlp commented:
I believe address of record is marked as dependent so that it can be displayed as a sub view on the form (otherwise it would only be available as a query combo box and all existing forms would have to be redone)

maxxxxxdlp commented:
Fixing https://github.com/specify/specify7/issues/114 would allow for editing independent resources in sub-views

grantfitzsimmons commented:
@melton-jason Is this something that just needs to be tested?

melton-jason commented:
Yes, testing can probably be done on this. There are still 2 cases I need further clarification/discussion on.

>  There is another concerning side-effect with this issue (currently in production). When Borrow is deleted, it tries to cascade and delete its associated Addressofrecord. If there are multiple records referencing the Addressofrecord, the error is thrown. However, if the Borrow is the only record referencing the Addressofrecord, both the Addressofrecord and Borrow will be deleted.
> This actually holds true for all of the following relationships:
> 
> Accession.addressofrecord
> Borrow.addressofrecord
> Exchangein.addressofrecord
> Exchangeout.addressofrecord
> Gift.addressofrecord
> Loan.addressofrecord
> Repositoryagreement.addressofrecord
>
> Interactions are flagged as dependents of Addressofrecord ..., but at a database level they do not require an Addressofrecord (similar to https://github.com/specify/specify7/issues/2299).
> @maxxxxxdlp or @grantfitzsimmons do you know why there is this discrepancy?

It has been awhile, so I would need to look into this specific issue again. Should deleting the interaction try and delete its address of record? I would immediately say no, because the address of record can be referenced by other interactions (or tables), hence the issue that caused #2511.

- - - 

(#1431)
> This will require some discussion. How should we handle LoanPreparation when it's associated Preparation is deleted? For comparison purposes to other branches I have set the relationship to null between Preparation and LoanPreparation. (The data seems to be preserved within the LoanPreparation, but it is 'orphaned' )
> 
> On other branches, LoanPreparations are a deletion blocker for Preparations, but the 'Trash' icon can still be clicked from the CollectionObject form and the user can 'attempt' to remove the Preparation. If we want to preserve the deletion blocker, we would need an implementation in the front end

We need to identify what _really_ needs to be done and what the issue is here. I would lean more towards this being a frontend issue with -to-many relationships. The issue is caused because a Preparation is removed from the Collection Object form, but that Preparation references a Loan Preparation, which causes the error to be thrown. 

@grantfitzsimmons For that view, I envision making any -to-many relationship like this display the warning symbol next to the trash can, which displays that resource's deletion blockers when clicked and prevents the deletion 
We can probably do a similar approach to the minus button in a subview representation of the resource.
Would this be an acceptable solution?

carlosmbe commented:
I'm going to request both UX and Dev Testing on this PR to conform with our new policy on needing 3 Reviews, Consisting of both Devs and UX people.

melton-jason committed:
Merge remote-tracking branch 'origin/production' into issue-1694

melton-jason committed:
Lint code with ESLint and Prettier

Triggered by 7d3cccaa35b511b7c865e4e772f7aaa159114d26 on branch refs/heads/issue-1694

melton-jason committed:
Return all instances of Additional Delete Blockers

melton-jason commented:
There were changes to the delete blocker dialog and the corresponding backend endpoint in `v7.9.0` (specifically in the following commits: https://github.com/specify/specify7/commit/dd2710154230cd219e44e0c1c4a0d5f036b50a6b, https://github.com/specify/specify7/commit/80087a29f4547c5c9fbd9ca37fd40309241af3fb), which resulted in the Additional Delete Blockers not being resolved correctly. 
This issue has since been resolved. 

To test this fix:
- Identify an Agent which is associated with a `Specifyuser`
  - Here is a Query which identifies all Agents in the `lsumz mammals` database: https://lsumzmammals-issue1694.test.specifysystems.org/specify/query/53/
-  The SpecifyUser should appear in the Agent's delete blocker dialog

pashiav reviewed (changes requested):
**https://github.com/specify/specify7/issues/1385** could not reproduce error on `edge` with https://fwri2023-edge.test.specifysystems.org/specify/query/48/

**https://github.com/specify/specify7/issues/1431** 
- Deletes preparation linked to interaction without error! 
- When I clicked **Save** for the loan, a new tab opened with this: [ReportException at _report_runner_run_.pdf](https://github.com/specify/specify7/files/14089804/ReportException.at._report_runner_run_.pdf). Is this related or a separate issue?


https://github.com/specify/specify7/assets/118556884/19864d7c-44bc-42ec-bcd6-ee2df43bc913



**https://github.com/specify/specify7/issues/2511** not fixed
[Specify 7 Crash Report - 2024-01-29T21_16_00.444Z.txt](https://github.com/specify/specify7/files/14089839/Specify.7.Crash.Report.-.2024-01-29T21_16_00.444Z.txt)
<img width="1467" alt="2511" src="https://github.com/specify/specify7/assets/118556884/9ef6ce5a-7064-4288-b91a-7afcb709ed4a">
https://herbrbge71423-issue1694.test.specifysystems.org/specify/view/borrow/1364/ 

**https://github.com/specify/specify7/pull/2806#issuecomment-1913346814** looks good!
<img width="1068" alt="Screenshot 2024-01-29 at 3 44 27 PM" src="https://github.com/specify/specify7/assets/118556884/37cb84cb-cb81-4dee-9caf-5456bba2a0cb">

melton-jason committed:
Merge branch 'production' into issue-1694
